[ADD] Add Smart Button in carbon factor form views

BonnetAdam opened the pull request.

MrFaBemol commented on onsp_co2/models/carbon_factor.py:
```diff
@@ -72,6 +81,21 @@ def _compute_child_qty(self):
         for factor in self:
             factor.child_qty = len(factor.child_ids)
 
+    @api.depends("chart_of_account_ids")           
+    def _compute_chart_of_account_qty(self):
+        for factor in self:
+            factor.chart_of_account_qty = len(factor.chart_of_account_ids)
+            
+    @api.depends("product_ids")           
+    def _compute_product_qty(self):
+        for factor in self:
+            factor.product_qty = len(factor.product_ids)
+            
+    @api.depends("product_categ_ids")           
+    def _compute_product_categ_qty(self):
+        for factor in self:
+            factor.product_categ_qty = len(factor.product_categ_ids)
```
Could you please use read_group as it is better performance wise?

You have 2 examples here: 
-  [OCA](https://github.com/OCA/project/blob/805cfc9366bc3ffb98ee90577e1738366ad4f09a/project_stock/models/project_task.py#L90-L96)
- [Odoo Enterprise](https://github.com/odoo/enterprise/blob/d0de5f1f5c6daa8537efacb62d73d2c8fd52c7f6/sign/models/res_partner.py#L10-L16)

BonnetAdam replied:
Fixed in the new commit

MrFaBemol commented on onsp_co2/views/carbon_factor.xml:
```diff
@@ -35,7 +35,55 @@
                         <div class="oe_button_box" name="button_box">
                             <button type="object" name="action_see_child_ids" class="oe_stat_button" icon="fa-sitemap">
-                                <field name="child_qty" widget="statinfo" string="Child Factors" />
+                                <div class="o_field_widget o_stat_info">
```
Is this worth for some plurals? :joy: 


Anyway, maybe change "Child Factors" and "Categorys"

BonnetAdam replied:
It was asked by BCH so i only execute the order xDDDDD

MrFaBemol commented on onsp_co2_purchase/__manifest__.py:
```diff
@@ -9,6 +9,7 @@
     "depends": [
         "onsp_co2",
         "purchase",
+        "contacts",
```
Why do you need contacts?

BonnetAdam replied:
Pretty sure so, the onsp_co2 used the model res.partner so i though we'll need it...

MrFaBemol replied:
res.partner is a model declared in the base module... :wink:

BonnetAdam replied:
Didn't know thanks for that... fix in https://github.com/MyCityCO2/mycityco2-engine/pull/13/commits/8b2e02ceb9c4ff9b92493a366cca5e58f01e1257

MrFaBemol commented on onsp_co2/models/carbon_factor.py:
```diff
@@ -25,6 +25,15 @@ class CarbonFactor(models.Model):
     child_qty = fields.Integer(compute="_compute_child_qty")
     value_ids = fields.One2many("carbon.factor.value", "factor_id")
+    
+    chart_of_account_ids = fields.One2many("account.account", "carbon_in_factor_id")
+    chart_of_account_qty = fields.Integer(compute="_compute_chart_of_account_qty")
+    
+    product_ids = fields.One2many("product.template", "carbon_in_factor_id")
+    product_qty = fields.Integer(compute="_compute_product_qty")
+    
+    product_categ_ids = fields.One2many("product.category", "carbon_in_factor_id")
+    product_categ_qty = fields.Integer(compute="_compute_product_categ_qty")
```
What about `out` factors?  You can't ignore them

BonnetAdam replied:
I totally forgot about them. In the new version I used read_group and also use the `out` factor

MrFaBemol reviewed (changes requested):
Nice QOL but please check my comments

Thanks!

MrFaBemol commented on onsp_co2/models/carbon_factor.py:
```diff
@@ -169,15 +254,13 @@ def get_value_infos_at_date(self, date=None) -> dict:
-        }
+        return self._generate_action(title="Child factors", model="carbon.factor")
+        
+    def action_see_chart_of_account_ids(self):
+        return self._generate_action(title="Chart of Account", model="account.account")
+
+    def action_see_product_ids(self):
+        return self._generate_action(title="Product", model="product.template")
+        
+    def action_see_product_categ_ids(self):
+        return self._generate_action(title="Product Category", model="product.category")
```
Could you add a translation wrapper (`_`) on titles please ?

BonnetAdam replied:
fix in https://github.com/MyCityCO2/mycityco2-engine/pull/13/commits/3d7e68249ab3bf6ec3e3740a7ece184995375887  and https://github.com/MyCityCO2/mycityco2-engine/pull/13/commits/80b7f8c3754de57495c6799f53501d28ec162ef6

MrFaBemol commented on onsp_co2/models/carbon_factor.py:
```diff
@@ -78,6 +88,81 @@ def _compute_descendant_ids(self):
+
+        Args:
+            model (str, optional): The name of the model to calculate the carbon factors for.
+
+        Returns:
+            dict: A dictionary where the keys are the carbon_in_factor_id and the values are the total count of carbon factors for that id.
+        """
+        data = self.env[model].read_group(
+            self._get_carbon_domain(), ["carbon_in_factor_id", "carbon_out_factor_id"], ["carbon_in_factor_id", "carbon_out_factor_id"], lazy=False
+        )
+        total_count = defaultdict(int)
+        for item in data:
+            total_count[item["carbon_in_factor_id"][0]] += item['__count']
+        
+        return total_count
+    
+    def _get_carbon_domain(self):
+        """
+        Generate the domain for carbon factor queries.
+
+        This function creates a domain that can be used in queries to filter data based on the carbon_in_factor_id and carbon_out_factor_id.
+
+        Returns:
+            list: A list representing the domain for carbon factor queries. The list includes a logical OR operator and two tuples, each specifying a field name and a condition.
+        """
+        return ['|', ("carbon_in_factor_id", "in", self.ids), ("carbon_out_factor_id", "in", self.ids)]
+    
+    def _generate_action(self, model: str, title: str):
+        """
+        Generate an action dictionary for the specified model and title.
+
+        This function creates an action dictionary that can be used to open a new window in the Odoo UI. The new window will display the specified model's data, filtered by the carbon domain.
+
+        Args:
+            model (str): The name of the model to display in the new window.
+            title (str): The title to display in the new window.
+
+        Returns:
+            dict: An action dictionary that can be used to open a new window in the Odoo UI.
+        """
```
Thanks for the docstrings!
Would be nice to use the python typing also like this:

```python
def _get_count_by_model(self, model: str) -> int:
```

BonnetAdam replied:
fix in https://github.com/MyCityCO2/mycityco2-engine/pull/13/commits/e2646e88af9a256804cde3923390095586be254c and https://github.com/MyCityCO2/mycityco2-engine/pull/13/commits/2355599f55bd61b03f599b65ef83206a77410c10

BonnetAdam committed:
[ADD] Add Smart Button in carbon factor form views

Add QOL button in the carbon factor view. Add function to get count easily using read_group
[FIX] Remove unused depends from onsp_co2_purchase

[FIX] Add typing in output of util function

[FIX] use of _ wrapper for string

[fix] small issues

[fix] forgot something

MrFaBemol commented:
@CrazyOutOff You can SMASH everything !

BonnetAdam commented:
@MrFaBemol it's been SMASHED !!!
